### ISSUE TITLE
references variable in templates

### ISSUE THREAD

**Issue by jgm** [jgm/pandoc#771]
Aditya Mahajan suggests on pandoc-discuss:

Is it possible to change the way bibliography is handled? Can you add
a $if(bibliography)$ check for the template file, and a $references$
tag for the list of references. Then, I can use the following in my
ConTeXt template:

```
$if(bibliography)$
%\subject{References}
\setupindenting[no]
\setupwhitespace[small]

$references$
$endif$
```

This may also circumvent the need to explicitly write `#References` in
the markdown file (but I don't know how other writers would handle
this).

**Comment by ousia** [comment]
> Experimented with implementing this, but found some rough edges. (This approach is difficult for epub, OPML, and some other formats.) My work is preserved in the branch issue771, but I think for now I'll stick to the present system.

@jgm, how about enabling a `# references` in a similar way to the proposals for TOC (#1612) and footnotes (#1720)?

This is much easier to use for the final user. And I think way more powerful.

**Comment by jgm** [comment]
Injecting the bibliography under a header with a special id (say, `references`) would indeed be easier technically than handling it with a template.

**Comment by ousia** [comment]
> Injecting the bibliography under a header with a special id (say, `references`) would indeed be easier technically than handling it with a template.

@jgm, excuse me, but I can’t refrain from asking: are header with special titles technically easier to implement than variables in templates?

**Comment by lierdakil** [comment]
Somewhat relevant pandoc-discuss thread: <https://groups.google.com/forum/#!topic/pandoc-discuss/HxmpFK-Ydus>
___
To answer @ousia, not necessarily, no. But for some particular cases, like epub, template is actually a *section* template, and not a *document* template (since epub is basically a collection of html pages). So this becomes a bit more difficult if handled with a template variable -- after all, we don't need references in *every* section.

**Comment by alick9188** [comment]
@lierdakil "after all, we don't need references in every section." It might not be the case where pandoc focuses on, but it is common for books to have references after each chapter. In latex chapterbib and biblatex can handle this.

**Comment by lierdakil** [comment]
Let me rephrase this then: we don't need ALL references in every section.
Because pandoc-citeproc does not handle the use-case you are talking about,
and it would be extremely hard to make it to.

And pandoc does have support for biblatex for latex/pdf output. And it's
completely separate feature, not discussed here.

2015-04-10 11:50 GMT+03:00 Alick Zhao <notifications@github.com>:

**Comment by sergiocorreia** [comment]
Since we are already using pandoc-citeproc to construct the references, can we just have another filter that deals with this issue, instead of having it as a Pandoc problem? EG:

`pandoc doc.md -F pandoc-citeproc -F pandoc-move-references`

Now, what would be the best approach? append the references after a header with `references` id, or is there a better alternative? (If it's the former, I already have a [sketch](https://www.pastiebin.com/58d43a58090cc) of a filter, from my pre-panflute days)

**Comment by slakkenhuis** [comment]
What @sergiocorreia suggests is similar to the approach I took when I ran into this problem. I made a `$references$` variable accessible from within the template (since I didn't need to export to EPUB or other formats where using a template would be problematic):

```haskell
import panflute as pf

def prepare(doc):
    doc.references = None

def action(el, doc):
    if isinstance(el, pf.Div) and el.identifier == "refs":
        doc.references = list(el.content)
        return []

def finalize(doc):
    if doc.references:
        doc.metadata["references"] = pf.MetaBlocks(*doc.references) 
    del doc.references

if __name__ == "__main__":
    pf.toJSONFilter(action, prepare, finalize)
```

This works fine, and I think inserting after specific headers is reasonable for a more general approach.

**Comment by jgm** [comment]
Pandoc-citeproc already allows you to insert a div with id refs, and in this case puts the references there.

Can this issue be closed, or is there something I'm missing?

**Comment by joundso** [comment]
> Pandoc-citeproc already allows you to insert a div with id refs, and in this case puts the references there.
> Can this issue be closed, or is there something I'm missing?

In case someone want's a quick copy & paste example:
```R
This is some text [@item1]

This is more text [@item2]

# References

<div id="refs"></div>

# appendix
```

Source: https://stackoverflow.com/a/44294306
